File: tasks/router.py
"""
Polls Telegram's getUpdates for any new message since the last check.
If it finds a recognized command, sets a GitHub Actions output so the
workflow knows whether to run the outreach job (and, later, ad/pin/
article jobs) this cycle.

Keeps track of the last processed Telegram update_id in a small state
file committed to the repo, so it doesn't reprocess the same command
on every 5-minute poll.

Usage: python tasks/router.py --check-only
"""

import os
import sys
import json
import urllib.request

REPO_ROOT = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
STATE_FILE = os.path.join(REPO_ROOT, "outreach", "telegram-state.json")

TELEGRAM_TOKEN = os.environ["TELEGRAM_BOT_TOKEN"]
API_BASE = f"https://api.telegram.org/bot{TELEGRAM_TOKEN}"

# command text (lowercased, stripped) -> github actions output flag to set
COMMAND_MAP = {
    "start outreach": "run_outreach",
    "/start_outreach": "run_outreach",
    # add more as new tasks come online, e.g.:
    # "make pins": "run_pinterest",
    # "write article": "run_medium",
}


def load_last_update_id():
    if os.path.exists(STATE_FILE):
        with open(STATE_FILE, "r") as f:
            return json.load(f).get("last_update_id", 0)
    return 0


def save_last_update_id(update_id):
    os.makedirs(os.path.dirname(STATE_FILE), exist_ok=True)
    with open(STATE_FILE, "w") as f:
        json.dump({"last_update_id": update_id}, f)


def get_updates(offset):
    url = f"{API_BASE}/getUpdates?offset={offset}&timeout=0"
    with urllib.request.urlopen(url, timeout=30) as resp:
        return json.loads(resp.read().decode())


def set_github_output(key, value):
    output_file = os.environ.get("GITHUB_OUTPUT")
    if output_file:
        with open(output_file, "a") as f:
            f.write(f"{key}={value}\n")
    else:
        print(f"[dry-run output] {key}={value}")


def main():
    last_id = load_last_update_id()
    data = get_updates(offset=last_id + 1)

    triggered_outputs = set()
    newest_id = last_id

    for update in data.get("result", []):
        newest_id = max(newest_id, update["update_id"])
        message = update.get("message", {})
        text = message.get("text", "").strip().lower()

        if text in COMMAND_MAP:
            output_key = COMMAND_MAP[text]
            triggered_outputs.add(output_key)
            print(f"Matched command '{text}' -> {output_key}")

    if newest_id != last_id:
        save_last_update_id(newest_id)
        # commit the state file so the next poll doesn't reprocess this
        os.system('git config user.name "github-actions"')
        os.system('git config user.email "actions@github.com"')
        os.system(f'git add {STATE_FILE}')
        os.system('git diff --staged --quiet || git commit -m "Update telegram poll state"')
        os.system('git push')

    for key in ["run_outreach", "run_pinterest", "run_medium", "run_fb_ads"]:
        set_github_output(key, "true" if key in triggered_outputs else "false")


if __name__ == "__main__":
    main()

File: tasks/router.ts
// tasks/router.ts
import { readFile, writeFile, mkdir } from 'fs/promises';
import { existsSync } from 'fs';
import path from 'path';
import { REPO_ROOT, OUTREACH_DIR } from './shared.js';

const STATE_FILE = path.join(OUTREACH_DIR, 'telegram-state.json');

const TELEGRAM_TOKEN = process.env.TELEGRAM_BOT_TOKEN;
if (!TELEGRAM_TOKEN) throw new Error('TELEGRAM_BOT_TOKEN not set');

const API_BASE = `https://api.telegram.org/bot${TELEGRAM_TOKEN}`;

// Command map – lowercased text -> GitHub Actions output flag
const COMMAND_MAP: Record<string, string> = {
  'start outreach': 'run_outreach',
  '/start_outreach': 'run_outreach',
  'make pins': 'run_pinterest',
  'write article': 'run_medium',
  'make ad': 'run_fb_ads',
  'quora drafts': 'run_quora',
};

async function loadLastUpdateId(): Promise<number> {
  if (!existsSync(STATE_FILE)) return 0;
  try {
    const data = JSON.parse(await readFile(STATE_FILE, 'utf-8'));
    return data.last_update_id || 0;
  } catch {
    return 0;
  }
}

async function saveLastUpdateId(updateId: number): Promise<void> {
  await mkdir(OUTREACH_DIR, { recursive: true });
  await writeFile(STATE_FILE, JSON.stringify({ last_update_id: updateId }, null, 2));
}

async function getUpdates(offset: number): Promise<any> {
  const url = `${API_BASE}/getUpdates?offset=${offset}&timeout=0`;
  const resp = await fetch(url, { timeout: 30000 });
  if (!resp.ok) throw new Error(await resp.text());
  return resp.json();
}

function setGitHubOutput(key: string, value: string): void {
  const outputFile = process.env.GITHUB_OUTPUT;
  if (outputFile) {
    // In Actions, we append to the file
    // We'll handle it asynchronously, but we need sync for this part – we can use fs.appendFileSync
    const fs = require('fs');
    fs.appendFileSync(outputFile, `${key}=${value}\n`);
  } else {
    console.log(`[dry-run] ${key}=${value}`);
  }
}

async function main() {
  const lastId = await loadLastUpdateId();
  const data = await getUpdates(lastId + 1);

  const triggeredOutputs = new Set<string>();
  let newestId = lastId;

  for (const update of data.result || []) {
    newestId = Math.max(newestId, update.update_id);
    const message = update.message || {};
    const text = (message.text || '').trim().toLowerCase();
    if (COMMAND_MAP[text]) {
      const flag = COMMAND_MAP[text];
      triggeredOutputs.add(flag);
      console.log(`Matched command '${text}' -> ${flag}`);
    }
  }

  if (newestId !== lastId) {
    await saveLastUpdateId(newestId);
    // Commit the state file – we'll let the workflow handle committing
    console.log('State file updated – workflow will commit it.');
  }

  // Set outputs for all possible flags
  const allFlags = ['run_outreach', 'run_pinterest', 'run_medium', 'run_fb_ads', 'run_quora'];
  for (const flag of allFlags) {
    setGitHubOutput(flag, triggeredOutputs.has(flag) ? 'true' : 'false');
  }
}

main().catch(err => {
  console.error(err);
  process.exit(1);
});
